trino/client.py

```diff
@@ -869,82 +877,179 @@ def map(self, value) -> Optional[float]:
+        hours = timezone_str[1:3]
+        minutes = timezone_str[4:6]
+        if timezone_str.startswith("-"):
+            return timezone(-timedelta(hours=int(hours), minutes=int(minutes)))
+        return timezone(timedelta(hours=int(hours), minutes=int(minutes)))
+    else:
+        return pytz.timezone(timezone_str)
+
+
+def _fraction_to_decimal(fraction_part: str) -> Decimal:
+    return Decimal(fraction_part or 0) / ROUNDING_FACTORS[len(fraction_part)]
+
+
+class TemporalType(Generic[PythonTemporalType], metaclass=abc.ABCMeta):
+    def __init__(self, time_part: PythonTemporalType, fraction_part: Decimal):
+        self._time_part = time_part
+        self._fraction_part = fraction_part
+
+    @abc.abstractmethod
+    def new_instance(self, value: PythonTemporalType, fraction: Decimal) -> TemporalType[PythonTemporalType]:
+        pass
+
+    @abc.abstractmethod
+    def to_python_type(self) -> PythonTemporalType:
+        pass
+
+    def round_to(self, precision: int) -> TemporalType:
+        """
+            Python datetime and time only support up to microsecond precision
+            In case the supplied value exceeds the specified precision,
+            the value needs to be rounded.
+        """
+        fraction_part = self._fraction_part
+        digits = abs(fraction_part.as_tuple().exponent)
+        if digits > precision:
+            rounding_factor = ROUNDING_FACTORS[precision]
+            rounded = fraction_part.quantize(Decimal(1 / rounding_factor))
+            if rounded == rounding_factor:
+                return self.new_instance(
+                    self.normalize(self.add_time_delta(timedelta(seconds=1))),
+                    Decimal(0)
+                )
+            return self.new_instance(self._time_part, rounded)
+        return self
+
+    @abc.abstractmethod
+    def add_time_delta(self, time_delta: timedelta) -> PythonTemporalType:
+        """
+            This method shall be overriden to implement fraction arithmetics.
+        """
+        pass
+
+    def normalize(self, value: PythonTemporalType) -> PythonTemporalType:
+        """
+            This method is overriden in `TimestampWithTimeZone` to normalize a datetime with a pytz timezone in case of
+            time part arithmetics (overflow of the fraction part).
+            This method is a no-op for datetime and time with numeric offsets.
+        """
+        return value
+
+
+class Time(TemporalType[time]):
+    def new_instance(self, value: time, fraction: Decimal) -> TemporalType[time]:
+        return Time(value, fraction)
+
+    def to_python_type(self) -> time:
+        if self._fraction_part > 0:
+            time_delta = timedelta(microseconds=int(self._fraction_part * ROUNDING_FACTORS[6]))
+            return self.add_time_delta(time_delta)
+        return self._time_part
+
+    def add_time_delta(self, time_delta: timedelta) -> time:
+        time_delta_added = datetime.combine(datetime(1, 1, 1), self._time_part) + time_delta
```

Comment:
Correct. See https://docs.python.org/3/library/datetime.html#determining-if-an-object-is-aware-or-naive.

Note that datetime.now() is also naive, it could also be used.